presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PreconditionRules.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import static com.facebook.presto.sql.planner.plan.Patterns.exchange;
+
+public class PreconditionRules
+{
+    private PreconditionRules()
+    {
+    }
+
+    public static Rule<ExchangeNode> checkRulesAreFiredBeforeAddExchangesRule()
```

Comment:
Hm... This somewhat is an attribute of a rule, and not actual invocation of a rule.
Ideally, the rule would match what it wants (but if it ever sees an exchange, it fails).

Having this as a separate method does not safeguard against putting a rule after `AddExchanges`, so this brings little more power than ordinary `// must be run before AddExchanges` comment.

I would prefer adding this to a rules returned from `PickTableLayout.rules()` and going with ordinary comment for "solitaire" rules.